helix-rest/src/main/java/org/apache/helix/rest/metadatastore/concurrency/ZkDistributedLeaderElection.java

```diff
@@ -0,0 +1,142 @@
+  private ZNRecord _currentLeaderInfo;
+
+  private String _myEphemeralSequentialPath;
+  private volatile boolean _isLeader;
+
+  public ZkDistributedLeaderElection(HelixZkClient zkClient, String basePath,
+      ZNRecord participantInfo) {
+    synchronized (this) {
+      if (zkClient == null || zkClient.isClosed()) {
+        throw new IllegalArgumentException("ZkClient cannot be null or closed!");
+      }
+      _zkClient = zkClient;
+      _zkClient.setZkSerializer(new ZNRecordSerializer());
+      if (basePath == null || basePath.isEmpty()) {
+        throw new IllegalArgumentException("lockBasePath cannot be null or empty!");
+      }
+      _basePath = basePath;
+      _participantInfo = participantInfo;
+      _isLeader = false;
+    }
+    init();
+  }
+
+  /**
+   * Create the base path if it doesn't exist and create an ephemeral sequential ZNode.
+   */
+  private void init() {
+    try {
+      _zkClient.createPersistent(_basePath, true);
+    } catch (ZkNodeExistsException e) {
+      // Okay if it exists already
+    }
+
+    // Create my ephemeral sequential node with my information
+    _myEphemeralSequentialPath = _zkClient
+        .create(_basePath + "/" + PREFIX, _participantInfo, CreateMode.EPHEMERAL_SEQUENTIAL);
+    if (_myEphemeralSequentialPath == null) {
+      throw new IllegalStateException(
+          "Unable to create ephemeral sequential node at path: " + _basePath);
+    }
+    tryAcquiringLeadership();
+  }
+
+  private void tryAcquiringLeadership() {
+    List<String> children = _zkClient.getChildren(_basePath);
+    Collections.sort(children);
+    String leaderName = children.get(0);
+    ZNRecord leaderInfo = _zkClient.readData(_basePath + "/" + leaderName, true);
+
+    String[] myNameArray = _myEphemeralSequentialPath.split("/");
+    String myName = myNameArray[myNameArray.length - 1];
+
+    if (leaderName.equals(myName)) {
+      // My turn for leadership
+      _isLeader = true;
+      _currentLeaderInfo = leaderInfo;
+      LOG.info("{} acquired leadership! Info: {}", myName, leaderInfo);
+    } else {
+      // Watch the ephemeral ZNode before me for a deletion event
+      String beforeMe = children.get(children.indexOf(myName) - 1);
+      _zkClient.subscribeDataChanges(_basePath + "/" + beforeMe, this);
+    }
+  }
+
+  public synchronized boolean isLeader() {
+    return _isLeader;
+  }
+
+  public synchronized ZNRecord getCurrentLeaderInfo() {
+    return _currentLeaderInfo;
+  }
+
+  @Override
+  public synchronized void handleStateChanged(Watcher.Event.KeeperState state) {
+    if (state == Watcher.Event.KeeperState.SyncConnected) {
+      init();
+    }
+  }
+
+  @Override
+  public void handleNewSession(String sessionId) {
+    return;
+  }
+
+  @Override
+  public void handleSessionEstablishmentError(Throwable error) {
+    return;
+  }
+
+  @Override
+  public void handleDataChange(String s, Object o) {
+    return;
+  }
+
+  @Override
+  public void handleDataDeleted(String s) {
```

Comment:
Was data delete referring the content of ZNode instead of ZNode it self?

Reply:
I'm not sure what you're asking about exactly...

>   private void fireDataChangedEvents(final String path, Set<IZkDataListenerEntry> listeners,
>       final OptionalLong notificationTime) {
>     try {
>       final ZkPathStatRecord pathStatRecord = new ZkPathStatRecord(path);
>       // Trigger listener callbacks
>       for (final IZkDataListenerEntry listener : listeners) {
>         _eventThread.send(new ZkEvent("Data of " + path + " changed sent to "
>             + listener.getDataListener() + " prefetch data: " + listener.isPrefetchData()) {
>           @Override
>           public void run() throws Exception {
>             // Reinstall watch before listener callbacks to check the znode status
>             if (!pathStatRecord.pathChecked()) {
>               pathStatRecord.recordPathStat(getStat(path, true), notificationTime);
>             }
>             if (!pathStatRecord.pathExists()) {
>               // no znode found at the path, trigger data deleted handler.
>               listener.getDataListener().handleDataDeleted(path);
>             } else {
>               Object data = null;
>               if (listener.isPrefetchData()) {
>                 if (LOG.isDebugEnabled()) {
>                   LOG.debug("Prefetch data for path: {}", path);
>                 }
>                 try {
>                   data = readData(path, null, true);
>                 } catch (ZkNoNodeException e) {
>                   LOG.warn("Prefetch data for path: {} failed.", path, e);
>                   listener.getDataListener().handleDataDeleted(path);
>                   return;
>                 }
>               }
>               listener.getDataListener().handleDataChange(path, data);
>             }
>           }
>         });
>       }

Does this help?